### PR TITLE
Python tracing module

### PR DESCRIPTION
**Pull request by MBaczun**
Adds a python module for tracing using zipkin and gRPC.

## Examples
1. cd to the example in `./function-images/tests/tracing/python/integ-tests/client-server`
2. Make the server and client images:
   ```bash
   make all-image
   ```
3. Run the docker-compose:
   ```bash
   docker-compose up --exit-code-from client
   ```
4. (optionally) Also try out the go client:
   ```bash
   docker-compose up --file ./docker-compose-goclient.yml up --exit-code-from client
   ```
If you want to run the example manually instead of using docker compose you can also try starting a zipkin container with `docker run -d -p 9411:9411 openzipkin/zipkin`, then starting the server in one terminal using `python ./server.py --zipkin http://localhost:9411/api/v2/spans` and start the client in a second terminal with `python ./greeter_client.py --zipkin http://localhost:9411/api/v2/spans`. You can view the resulting traces in your browser at `http://localhost:9411/zipkin/`.
      
## Sample output
timeline
graph

### CODE REVIEW

File: utils/tracing/python/examples/client-server/greeter_client.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+            if len(sys.argv) == 2:
+                if sys.argv[1] == 'rec':
+                    name = 'record'
+                elif sys.argv[1] == 'rep':
+                    name = 'replay'
+                else:
+                    exit(-1)
+
+            response = stub.SayHello(helloworld_pb2.HelloRequest(name=name))
```
Comment (ustiugov) [inline utils/tracing/python/examples/client-server/greeter_client.py:54]: simplify
```suggestion
            response = stub.SayHello(helloworld_pb2.HelloRequest(name="world"))
```

File: utils/tracing/python/examples/client-server/requirements.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+futures
```
Comment (ustiugov) [inline utils/tracing/python/examples/client-server/requirements.txt:1]: you probably need to install opentelemetry as well, right?

File: utils/tracing/python/examples/client-server/server.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    def SayHello(self, request, context):
+        if request.name == "record":
+            msg = 'Hello, %s!' % responses[0]
+        elif request.name == "replay":
+            msg = 'Hello, %s!' % responses[1]
+        else:
+            msg = 'Hello, %s!' % request.name
+
+        return helloworld_pb2.HelloReply(message=msg)
```
Comment (ustiugov) [inline utils/tracing/python/examples/client-server/server.py:48]: simplify
```suggestion
        return helloworld_pb2.HelloReply(message= request.name)
```

File: utils/tracing/python/examples/spans/spanExample.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+from ... import tracing
```
Comment (ustiugov) [inline utils/tracing/python/examples/spans/spanExample.py:1]: replace with a unit test

**Review by ustiugov — CHANGES_REQUESTED** [review]
- [ ] move the client-server sources to function-images/tests/tracing/python
- [ ] add a CI test 9 running them with plain Python is good enough
- [ ] add a unit test for Python tracing
- [ ] make sure opentelemetry dependencies are installed properly (e.g., via the requirements.txt file)
- [ ] add a screenshot from Zipkin
- [ ] add copyrights
- [ ] remove empty files unless they are required

File: .github/workflows/python_tracing.yml [hunk 5]
```diff
@@ -0,0 +1,29 @@
+name: Python tracing CI
```
Comment (ustiugov) [inline .github/workflows/python_tracing.yml:1]: add this to the unit-test files outside of the matric though. too many workflows as of now

File: function-images/tests/tracing/python/client-server/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+#STOPSIGNAL SIGKILL
+
+#ENTRYPOINT ["/usr/local/bin/python", "/server.py"]
+
+FROM python:3.8-slim-buster as server
+COPY ./function-images/tests/tracing/python/client-server/requirements.txt .
+COPY ./function-images/tests/tracing/python/client-server/*.py /
+COPY ./utils/tracing/python/tracing.py /
+RUN pip3 install --user -r requirements.txt
+
```
Comment (ustiugov) [inline function-images/tests/tracing/python/client-server/Dockerfile:47]: create WORKDIR, e.g., /app and copy the stuff there. don't work in the `/` folder, calls for trouble.

File: function-images/tests/tracing/python/client-server/Makefile [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+all-image: server-image client-image go-client-image
+
+server-image: Dockerfile server.py helloworld_pb2_grpc.py helloworld_pb2.py
+	cd ../../../../../ && docker build --tag vhiveease/py-tracing-server:latest --target server -f ./function-images/tests/tracing/python/client-server/Dockerfile .
```
Comment (ustiugov) [inline function-images/tests/tracing/python/client-server/Makefile:26]: define an env variable and use it instead:
```
ROOT=../../../../../
```
watch the quote marks, not sure about those.

File: function-images/tests/tracing/python/client-server/goclient/goclient.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	})
+	log.SetOutput(os.Stdout)
+
+	shutdown, err := tracing.InitBasicTracer(*url, "go client")
+	if err != nil {
+		log.Warn(err)
+	}
+	defer shutdown()
+
+	conn, err := grpc.Dial(fmt.Sprintf("%v:%v", *address, *clientPort), grpc.WithInsecure(), grpc.WithUnaryInterceptor(otelgrpc.UnaryClientInterceptor()))
```
Comment (ustiugov) [inline function-images/tests/tracing/python/client-server/goclient/goclient.go:63]: line is too long, 80-100 char max

File: function-images/tests/tracing/python/client-server/requirements.txt [hunk 9]
```diff
@@ -0,0 +1,10 @@
+futures
+grpcio
```
Comment (ustiugov) [inline function-images/tests/tracing/python/client-server/requirements.txt:2]: fix this one as well

**Review by ustiugov — CHANGES_REQUESTED** [review]
hope this helps

- [x] rename tracing/python/client-server to  tracing/python/integ-tests/client-server
- [x] add a line to the changelog (in the UNRELEASED section)

File: function-images/tests/tracing/python/client-server/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+#COPY ./utils/tracing/python/tracing.py /
+#COPY ./function-images/tests/tracing/python/client-server/*.py /
+#COPY --from=builder_workload /root/.local /root/.local
+#RUN apk add libstdc++ --update --no-cache
+
+#EXPOSE 50051
+
+#STOPSIGNAL SIGKILL
+
+#ENTRYPOINT ["/usr/local/bin/python", "/server.py"]
```
Comment (ustiugov) [inline function-images/tests/tracing/python/client-server/Dockerfile:40]: remove

**Review by ustiugov — APPROVED** [review]
double check that you get meaningful traces after all the changes. then merge it

**Review by ustiugov — CHANGES_REQUESTED** [review]
rebase please!

**Comment by MBaczun** [comment]
Zipkin traces still all look correct 👍

**Review by ustiugov — APPROVED** [review]

**Pull request merged**